### PR TITLE
add support for merging relationship properties

### PR DESCRIPTION
**Pull request by ajtmccarty**
IFC-675

more unit tests
support for adding/deleting/updating relationship properties in a merge

there is some special handling required for updates to a cardinality=one relationship where IS_PROTECTED or IS_VISIBLE are NOT updated b/c the diff does not record changes to the IS_PROTECTED or IS_VISIBLE property(ies) in that case, BUT we still need to update those edges in the database to be active on main. the unit tests indicate that my logic works for this case, but there is probably a better way to handle it

still need to add integration tests

### CODE REVIEW

File: backend/infrahub/core/diff/merger/serializer.py [hunk 1]
```diff
@@ -169,38 +172,105 @@ def _serialize_attribute(
...
         )
         return attr_dict, attr_prop_dict
 
     def _serialize_relationship_element(
-        self, relationship_diff: EnrichedDiffSingleRelationship, relationship_identifier: str
+        self, relationship_diff: EnrichedDiffSingleRelationship, relationship_identifier: str, node_uuid: str
     ) -> tuple[list[RelationshipMergeDict], list[RelationshipPropertyMergeDict]]:
         relationship_dicts = []
+        # start with default values for IS_VISIBLE and IS_PROTECTED b/c we always want to update them during a merge
+        added_property_dicts: dict[DatabaseEdgeType, PropertyMergeDict] = {
```
Comment (ajtmccarty) [inline backend/infrahub/core/diff/merger/serializer.py:201]: these are initialized with IS_PROTECTED and IS_VISIBLE starting values to make sure that they are always updated for a given relationship

File: backend/infrahub/core/diff/query/merge.py [hunk 2]
```diff
@@ -289,43 +308,80 @@ async def query_init(self, db: InfrahubDatabase, **kwargs: Any) -> None:
...
+                WITH attr_rel, property_diff
+                OPTIONAL MATCH (attr_rel)-[r_vis_pro]->(bool:Boolean)
+                WHERE property_diff.property_type IN ["IS_VISIBLE", "IS_PROTECTED"]
+                AND r_vis_pro.branch IN [$source_branch, $target_branch]
+                AND type(r_vis_pro) = property_diff.property_type
+                AND (property_diff.value IS NULL OR bool.value = property_diff.value)
+                RETURN bool
+                ORDER BY r_vis_pro.from DESC
+                LIMIT 1
+            }
```
Comment (ajtmccarty) [inline backend/infrahub/core/diff/query/merge.py:368]: if we don't have a specific value for IS_VISIBLE or IS_PROTECTED, we will get the latest value on the source branch and use that

File: backend/infrahub/core/diff/merger/serializer.py [hunk 3]
```diff
@@ -104,7 +104,8 @@ async def serialize_diff(
                 attribute_diff, attribute_property_diff = self._serialize_attribute(
                     attribute_diff=attr_diff, node_uuid=node.uuid, node_kind=node.kind
                 )
-                attribute_diffs.append(attribute_diff)
+                if attribute_diff:
+                    attribute_diffs.append(attribute_diff)
```
Comment (ogenstad) [inline backend/infrahub/core/diff/merger/serializer.py:108]: Not related to this PR but in this context I think it's a bit confusing to have both `attr_diff` and `attribute_diff` unless there's something I'm missing I think we should have an issue to choose other names in the future for these two variables or at least one of them.

**Review by ogenstad — APPROVED** [review]

File: backend/infrahub/core/diff/merger/serializer.py (hunk 3 above)
Comment (ajtmccarty) [reply backend/infrahub/core/diff/merger/serializer.py:108]: agreed. that is a bad name. will update it in my next PR

**Pull request merged**